orchagent/aclorch.cpp

```diff
@@ -5112,3 +5658,49 @@ void AclOrch::removeAllAclRuleStatus()
     }
 }
 
+MetaDataMgr::MetaDataMgr()
+{
+    for (uint8_t i = METADATA_VALUE_START; i <= METADATA_VALUE_END; i++)
+    {
+        m_freeMetadata.push_back(i);
+    }
+}
+uint8_t MetaDataMgr::getFreeMetaData(uint8_t dscp)
+{
+    uint8_t metadata =METADATA_VALUE_INVALID;
+    if (m_dscpMetadata.find(dscp) != m_dscpMetadata.end())
+    {
+        // dscp value has a metadata value assigned to it.
+        metadata = m_dscpMetadata[dscp];
+    }
+    else
+    {
+        if (m_freeMetadata.empty())
+        {
+            SWSS_LOG_ERROR("Metadata Value not available for allocation.");
+            return metadata;
+        }
+        metadata = m_freeMetadata.front();
+        m_freeMetadata.erase(m_freeMetadata.begin());
+        m_dscpMetadata[dscp] = metadata;
+    }
+    m_MetadataRef[metadata] += 1;
+    return metadata;
+}
+
+void MetaDataMgr::recycleMetaData(uint8_t metadata)
+{
+    m_MetadataRef[metadata] -= 1;
+    if (m_MetadataRef[metadata] == 0)
+    {
+        for (auto iter = m_dscpMetadata.begin(); iter != m_dscpMetadata.end(); ++iter)
+        {
```

Comment:
Remove an item from map in a for loop when the iter is being referenced can lead to crash.

Reply:
good catch, ill add more tests for this.

Reply:
added the fix and. added a test

Reply:
Actually there is an easier way to do this. `iter = m_dscpMetadata.erase(dscpToErase)`

Reply:
i am not sure what you mean here. the metadataMgr has a map of dscp -> metada **m_dscpMetadata** and refcount of metadata. when metadata refcount becomes 0, the logic finds goves over the metadataMgr to find the dscp which as this metadata asosciated. and finally removes it.
how is `iter = m_dscpMetadata.erase(dscpToErase)` easier. could you please explain.

Reply:
Below is an example.
```
for (auto it = m.cbegin(); it != m.cend() /* not hoisted */; /* no increment */)
{
  if (must_delete)
  {
    m.erase(it++);    // or "it = m.erase(it)" since C++11
  }
  else
  {
    ++it;
  }
}
```